spec/client/ics-007-tendermint-client/README.md

```diff
@@ -206,44 +206,53 @@ Tendermint client validity checking uses the bisection algorithm described in th
 
 ```typescript
 function verifyClientMessage(
-  clientMsg: ClientMessage) {
+  clientMsg: ClientMessage) => bool {
     switch typeof(clientMsg) {
       case Header:
-        verifyHeader(clientMsg)
+        return verifyHeader(clientMsg)
       case Misbehaviour:
-        verifyHeader(clientMsg.h1)
-        verifyHeader(clientMsg.h2)
+        return verifyHeader(clientMsg.h1) && verifyHeader(clientMsg.h2)
     }
 }
 ```
 
 Verify validity of regular update to the Tendermint client
 
 ```typescript
-function verifyHeader(header: Header) {
+function verifyHeader(header: Header) => bool {
     clientState = get("clients/{header.identifier}/clientState")
-    // assert trusting period has not yet passed
-    assert(currentTimestamp() - clientState.latestTimestamp < clientState.trustingPeriod)
-    // assert header timestamp is less than trust period in the future. This should be resolved with an intermediate header.
-    assert(header.timestamp - clientState.latestTimeStamp < trustingPeriod)
+    // trusting period must have not yet passed
+    if clientState.latestTimestamp + clientState.trustingPeriod > currentTimestamp() {
+      return false
+    }
+    // header timestamp must be less than or equal to the trusting period in the future. This should be resolved with an intermediate header.
+    if clientState.latestTimeStamp + clientState.trustingPeriod > header.timestamp {
```

Comment:
Cannot find this check in the implementation.

Reply:
Good point. I also cannot find it...

Reply:
So the tendermint light client verification function actually does the following check:

`clientState.latestTimestamp + clientState.trustingPeriod <= header.Timestamp + clientState.MaxClockDrift`

By the following checks:

https://github.com/tendermint/tendermint/blob/af2981a2f75500bc02412bad590fd26228cf5bc3/light/verifier.go#L207

https://github.com/tendermint/tendermint/blob/af2981a2f75500bc02412bad590fd26228cf5bc3/light/verifier.go#L176

Reply:
The check 

`currentTimestamp() >= clientState.latestTimestamp + clientState.trustingPeriod`

is done [here](https://github.com/tendermint/tendermint/blob/af2981a2f75500bc02412bad590fd26228cf5bc3/light/verifier.go#L46).

But looks like this check

`header.timestamp >= clientState.latestTimeStamp + clientState.trustingPeriod`

is not in the implementation.

And the implementation has these two checks, that we don't seem to have in the spec:

https://github.com/tendermint/tendermint/blob/af2981a2f75500bc02412bad590fd26228cf5bc3/light/verifier.go#L170 (which I think would translate to `header.timestamp <= clientState.latestTimestamp`)

https://github.com/tendermint/tendermint/blob/af2981a2f75500bc02412bad590fd26228cf5bc3/light/verifier.go#L176 (which I think would translate to `header.timestamp >= currentTimestamp() + clientState.maxClockDrift`)

Reply:
> https://github.com/tendermint/tendermint/blob/af2981a2f75500bc02412bad590fd26228cf5bc3/light/verifier.go#L170 (which I think would translate to header.timestamp <= clientState.latestTimestamp)

No because the trusted consensus state is not necessarily the latest one.

My point on the two links above.

The first checks the following:

`now < trustedConsState.timestamp + trustingPeriod`

And the second checks:

`now > header.timestamp + maxClockDrift`

Putting these together yields:

`header.timeStamp + maxClockDrift < trustedConsensusState.timestamp + clientState.trustingPeriod`

The highest trusted consensus state is the client state's latest timestamp, so we can replace this in the future update case with:

`header.timeStamp + maxClockDrift < clientState.latesttimestamp + clientState.TrustingPeriod`

Which would require the check:

`header.timestamp >= consensusState.timestamp  + clientState.TrustingPeriod - clientState.maxClockDrift`

So i would recommend replacing the line with the above conditional check